### PR TITLE
issue-1010 added ApiKeyResolver to DefaultClientBuilder, DefaultClien…

### PR DESCRIPTION
**Pull request by iancorcoran**
…t and DefaultDataStore

### CODE REVIEW

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy [hunk 1]
```diff
@@ -90,22 +93,40 @@ class DefaultClientTest {
...
     }
 
     @Test
+    void testConstructorApiKeyResolverNull() {
+
+        String baseUrl = "http://localhost:8080/v1"
+        def proxy = createMock(com.stormpath.sdk.client.Proxy)
+        def cacheManager = createMock(CacheManager)
+        def authcScheme = AuthenticationScheme.SAUTHC1
+        def clientCredentials = createNiceMock(ClientCredentials)
```
Comment (mraible) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy]: Is there a particular reason you're using a nice mock here? We generally use strict mocks so we can verify the order methods are called on the mock.

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy [hunk 2]
```diff
@@ -90,22 +93,40 @@ class DefaultClientTest {
...
+        } catch (IllegalArgumentException ex) {
+            assertEquals(ex.getMessage(), "apiKeyResolver argument cannot be null.")
+        }
+    }
+
+    @Test
     void testGetDataStore() {
 
         def apiKeyCredentials = createNiceMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
Comment (mraible) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy]: strict mock

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy [hunk 3]
```diff
@@ -115,6 +136,7 @@ class DefaultClientTest {
     @Test
     void testDirtyPropertiesNotSharedAmongDifferentResourcesWithSameHref() {
         def apiKeyCredentials = createNiceMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
Comment (mraible) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy]: Strict mock preferred

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java [hunk 4]
```diff
@@ -423,6 +423,19 @@ public LocaleChangeInterceptor stormpathLocaleChangeInterceptor() {
...
+        return super.stormpathRequestClientAttributeNames();
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(name = "stormpathRequestApplicationAttributeNames")
+    public Set<String> stormpathRequestApplicationAttributeNames() {
+        return super.stormpathRequestApplicationAttributeNames();
+    }
+
+
```
Comment (mraible) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java]: Remove extra line break.

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java [hunk 5]
```diff
@@ -423,6 +423,19 @@ public LocaleChangeInterceptor stormpathLocaleChangeInterceptor() {
     }
 
     @Bean
+    @ConditionalOnMissingBean(name = "stormpathRequestClientAttributeNames")
```
Comment (mraible) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java]: Are these new beans related to this PR?

File: impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java [hunk 6]
```diff
@@ -107,37 +107,38 @@
...
+    private final ApiKeyResolver apiKeyResolver;
 
     /**
      * @since 1.0.RC3
      */
     public static final String USER_AGENT_STRING = UserAgent.getUserAgentString();
 
-    public DefaultDataStore(RequestExecutor requestExecutor, ApiKeyCredentials apiKeyCredentials) {
-        this(requestExecutor, DEFAULT_API_VERSION, apiKeyCredentials);
+    public DefaultDataStore(RequestExecutor requestExecutor, ApiKeyCredentials apiKeyCredentials, ApiKeyResolver apiKeyResolver) {
```
Comment (mraible) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java]: Add `since` javadoc to these new constructors.

File: impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java [hunk 7]
```diff
@@ -242,6 +244,12 @@ public ClientBuilder setClientCredentials(ClientCredentials clientCredentials) {
         return this;
     }
 
+    public ClientBuilder setApiKeyResolver(ApiKeyResolver apiKeyResolver) {
```
Comment (mraible) [inline impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java]: Add `since` javadoc.

File: impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java [hunk 8]
```diff
@@ -58,6 +60,14 @@ public void setApiKeySecret(String apiKeySecret) {
         this.apiKeySecret = apiKeySecret;
     }
 
+    public ApiKeyResolver getApiKeyResolver() {
```
Comment (mraible) [inline impl/src/main/java/com/stormpath/sdk/impl/config/ClientConfiguration.java]: Add `since` javadoc to new methods.

File: impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy [hunk 9]
```diff
@@ -88,4 +91,68 @@ class DefaultClientBuilderTestCustomCredentials{
         assertEquals client.dataStore.apiKey.secret, secret
     }
 
+    @Test
+    void testCustomClientCredentialsRequireApiKeyResolver(){
```
Comment (mraible) [inline impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy]: We sometimes add `since` javadoc to new tests. Don't know if it's necessary, but doesn't hurt to be thorough!

File: impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy [hunk 10]
```diff
@@ -88,4 +91,68 @@ class DefaultClientBuilderTestCustomCredentials{
...
+        builder = new DefaultClientBuilder()
+        builder.setClientCredentials(customCredentials)
+        builder.setApiKeyResolver(apiKeyResolver)
+        def testClient = builder.build()
+
+        assertEquals testClient.dataStore.apiKey.id, keyId
+        assertEquals testClient.dataStore.apiKey.secret, keySecret
+    }
+
+
```
Comment (mraible) [inline impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientBuilderTest.groovy]: Remove extra line breaks.

File: impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy [hunk 11]
```diff
@@ -27,6 +28,7 @@ class DefaultClientTest extends PowerMockTestCase {
     void testCreateRequestExecutor() {
 
         def apiKeyCredentials = createMock(ApiKeyCredentials)
+        def apiKeyResolver = createNiceMock(ApiKeyResolver)
```
Comment (mraible) [inline impl/src/test/groovy/com/stormpath/sdk/impl/client/DefaultClientTest.groovy]: strict mock

File: impl/src/test/groovy/com/stormpath/sdk/impl/directory/DefaultCustomDataTest.groovy [hunk 12]
```diff
@@ -218,7 +219,7 @@ class DefaultDataStoreDelegateTo extends DefaultDataStore {
     private CustomData customData;
 
     DefaultDataStoreDelegateTo(RequestExecutor requestExecutor, CustomData customData) {
-        super(requestExecutor, "https://api.stormpath.com/v1", (ApiKeyCredentials)createNiceMock(ApiKeyCredentials))
+        super(requestExecutor, "https://api.stormpath.com/v1", (ApiKeyCredentials)createNiceMock(ApiKeyCredentials), createNiceMock(ApiKeyResolver))
```
Comment (mraible) [inline impl/src/test/groovy/com/stormpath/sdk/impl/directory/DefaultCustomDataTest.groovy]: I'd suggest `createStrictMock`, but I guess nice mock is used for `ApiKeyCredentials`, so I'd simply suggest consistency - which you have here.

File: impl/src/test/groovy/com/stormpath/sdk/impl/ds/DefaultDataStoreTest.groovy [hunk 13]
```diff
@@ -390,4 +403,20 @@ class DefaultDataStoreTest {
         }
     }
 
+    @Test
+    void testApiKeyResolverReturnsCorrectApiKey(){
+        def requestExecutor = createNiceMock(RequestExecutor)
+        def apiKeyForCredentials = createNiceMock(ApiKey)
+        def apiKeyForResolver = createNiceMock(ApiKey)
```
Comment (mraible) [inline impl/src/test/groovy/com/stormpath/sdk/impl/ds/DefaultDataStoreTest.groovy]: I'd suggest strict over nice mocks.

**Comment by mraible** [comment]
Reviewed - back to you @iancorcoran.

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java (hunk 5 above)
Comment (iancorcoran) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java]: These were part of an earlier PR of mine that went in with 1.0.4. Looks like @jarias may have accidentally overwritten them when merging the 1.1 branch into master. Not strictly related to adding an ApiKeyResolver I guess.

File: impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java (hunk 6 above)
Comment (iancorcoran) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/DefaultDataStore.java]: I don't mind adding these but as a rule this doesn't seem to be done in the impl module outside of class level javadoc?

**Comment by iancorcoran** [comment]
@mraible Believe I've addressed your comments outside of the open javadoc question. In regards to your question around whether the DefaultApiKeyResolver should be added as a property that can be overridden... we have an internal use case where we need to supply our own ApiKeyResolver (ping me on slack if you want to discuss more on the details of it). Similar to supplying custom ClientCredentials to the ClientBuilder, it is not really applicable outside of our internal use case, and follows the same pattern of just being exposed on the DefaultClientBuilder rather than the api ClientBuilder interface.
